File: dromozoa/compiler/runtime/runtime_cxx.lua
return [[
#ifndef DROMOZOA_COMPILER_RUNTIME_CXX_HPP
#define DROMOZOA_COMPILER_RUNTIME_CXX_HPP

#include <cmath>
#include <cstddef>
#include <cstdint>
#include <cstring>
#include <functional>
#include <initializer_list>
#include <iomanip>
#include <iostream>
#include <map>
#include <memory>
#include <sstream>
#include <stdexcept>
#include <string>
#include <tuple>
#include <utility>
#include <vector>

namespace dromozoa {
  namespace runtime {
    enum struct type_t : std::uint8_t {
      nil,
      boolean,
      number,
      string,
      table,
      function,
    };

    class error_t : public std::runtime_error {
    public:
      error_t(const char* message)
        : std::runtime_error(message) {}
    };

    class value_t;
    using array_t = std::vector<value_t>;
    using array_ptr = std::shared_ptr<array_t>;
    using upvalues_t = std::vector<std::tuple<array_ptr, std::size_t>>;
    using upvalues_ptr = std::shared_ptr<upvalues_t>;
    using string_t = std::string;
    using string_ptr = std::shared_ptr<string_t>;
    class table_t;
    using table_ptr = std::shared_ptr<table_t>;
    class function_t;
    using function_ptr = std::shared_ptr<function_t>;

    const value_t& nil() noexcept;

    array_ptr newarray(std::initializer_list<value_t> values, array_ptr array = nullptr);
    array_ptr newarray2(const value_t& value, array_ptr array);

    inline const value_t& get(array_ptr array, std::size_t index) noexcept {
      if (array && index < array->size()) {
        return (*array)[index];
      } else {
        return nil();
      }
    }

    class table_t {
      friend class value_t;
    public:
      using map_t = std::map<value_t, value_t>;
      const value_t& getmetafield(const char* event) const noexcept;
    private:
      map_t map_;
      table_ptr metatable_;
    };

    class function_t {
    public:
      using closure_t = std::function<array_ptr(array_ptr, array_ptr)>;
      template <class T>
      function_t(std::size_t argc, bool vararg, T&& closure)
        : argc_(argc), vararg_(vararg), closure_(std::forward<T>(closure)) {}
      array_ptr call(array_ptr array) const;
    private:
      std::size_t argc_;
      bool vararg_;
      closure_t closure_;
    };

    class value_t {
    public:
      value_t() noexcept {
        type_ = type_t::nil;
      }

      value_t(const value_t& that) noexcept {
        construct(that);
      }

      value_t(value_t&& that) noexcept {
        construct(that);
      }

      value_t& operator=(const value_t& that) noexcept {
        destruct();
        construct(that);
        return *this;
      }

      value_t& operator=(value_t&& that) noexcept {
        destruct();
        construct(that);
        return *this;
      }

      ~value_t() noexcept {
        destruct();
      }

      friend value_t boolean(bool boolean);
      friend value_t number(double number);
      friend value_t string(const char* data, std::size_t size);
      friend value_t string(const std::string& data);
      friend value_t table();
      template <class T>
      friend value_t function(std::size_t argc, bool vararg, T&& closure);

      bool is_nil() const noexcept {
        return type_ == type_t::nil;
      }

      bool is_boolean() const noexcept {
        return type_ == type_t::boolean;
      }

      bool is_number() const noexcept {
        return type_ == type_t::number;
      }

      bool is_string() const noexcept {
        return type_ == type_t::string;
      }

      bool is_table() const noexcept {
        return type_ == type_t::table;
      }

      bool is_function() const noexcept {
        return type_ == type_t::function;
      }

      const value_t& getmetafield(const char* event) const noexcept {
        if (!is_table()) {
          return nil();
        }
        return table_->getmetafield(event);
      }

      table_ptr getmetatable() const noexcept {
        if (!is_table()) {
          return nullptr;
        }
        return table_->metatable_;
      }

      void setmetatable(const value_t& metatable) {
        if (!is_table()) {
          throw error_t("table expected");
        }
        if (!metatable.is_nil() && !metatable.is_table()) {
          throw error_t("nil or table expected");
        }
        if (table_->metatable_) {
          if (!table_->metatable_->getmetafield("__metatable").is_nil()) {
            throw error_t("cannot change a protected metatable");
          }
        }
        if (metatable.is_nil()) {
          table_->metatable_ = nullptr;
        } else {
          table_->metatable_ = metatable.table_;
        }
      }

      array_ptr call(array_ptr array) const {
        if (is_function()) {
          return function_->call(array);
        }
        const auto& field = getmetafield("__call");
        if (!field.is_function()) {
          throw error_t("attempt to call a non-function value");
        }
        return field.function_->call(newarray2(field, array));
      }

      void call0(array_ptr array) const {
        call(array);
      }

      value_t call1(array_ptr array = nullptr) const {
        return get(call(array), 0);
      }

      value_t gettable(const value_t& index) const {
        if (!is_table()) {
          throw error_t("table expected");
        }
        const auto i = table_->map_.find(index);
        if (i != table_->map_.end()) {
          return i->second;
        }
        const auto& field = getmetafield("__index");
        if (!field.is_nil()) {
          if (field.is_function()) {
            return field.call1(newarray({ field, *this, index }));
          } else {
            return field.gettable(index);
          }
        }
        return nil();
      }

      void settable(const value_t& index, const value_t& value) const {
        if (!is_table()) {
          throw error_t("table expected");
        }
        const auto i = table_->map_.find(index);
        if (i == table_->map_.end()) {
          const auto& field = getmetafield("__newindex");
          if (!field.is_nil()) {
            if (field.is_function()) {
              field.call0(newarray({ field, *this, index, value }));
              return;
            } else {
              field.settable(index, value);
              return;
            }
          }
        }
        if (value.is_nil()) {
          table_->map_.erase(index);
        } else {
          table_->map_[index] = value;
        }
      }

      void setlist(double index, const value_t& value) const;
      void setlist(double index, array_ptr array) const;

      std::string type() const {
        switch (type_) {
          case type_t::nil:
            return "nil";
          case type_t::boolean:
            return "boolean";
          case type_t::number:
            return "number";
          case type_t::string:
            return "string";
          case type_t::table:
            return "table";
          case type_t::function:
            return "function";
        }
        throw error_t("???");
      }

      std::string tostring() const {
        switch (type_) {
          case type_t::nil:
            return "nil";
          case type_t::boolean:
            if (boolean_) {
              return "true";
            } else {
              return "false";
            }
          case type_t::number:
            {
              std::ostringstream out;
              out << std::setprecision(17) << number_;
              return out.str();
            }
          case type_t::string:
            return *string_;
          case type_t::table:
            {
              const auto& field = getmetafield("__tostring");
              if (!field.is_nil()) {
                return field.call1(newarray({ *this })).tostring();
              }
              std::ostringstream out;
              out << "table: " << table_.get();
              return out.str();
            }
          case type_t::function:
            {
              std::ostringstream out;
              out << "function: " << function_.get();
              return out.str();
            }
        }
        throw error_t("!!!");
      }

      double tonumber() const {
        if (is_number()) {
          return number_;
        }
        throw error_t("number expected");
      }

      std::int64_t tointeger() const {
        return tonumber();
      }

      value_t len() const;

      bool lt(const value_t& that) const {
        if (is_number() && that.is_number()) {
          return number_ < that.number_;
        } else if (is_string() && that.is_string()) {
          return string_ < that.string_;
        }
        throw error_t("attempt to compare...");
      }

      bool le(const value_t& that) const {
        if (is_number() && that.is_number()) {
          return number_ <= that.number_;
        } else if (is_string() && that.is_string()) {
          return string_ <= that.string_;
        }
        throw error_t("attempt to compare...");
      }

      friend std::ostream& operator<<(std::ostream& out, const value_t& self) {
        return out << self.tostring();
      }

      bool operator==(const value_t& that) const noexcept {
        if (type_ != that.type_) {
          return false;
        }
        switch (type_) {
          case type_t::nil:
            return true;
          case type_t::boolean:
            return boolean_ == that.boolean_;
          case type_t::number:
            return number_ == that.number_;
          case type_t::string:
            return *string_ == *that.string_;
          case type_t::table:
            return table_ == that.table_;
          case type_t::function:
            return function_ == that.function_;
        }
      }

      bool operator!=(const value_t& that) const noexcept {
        return !operator==(that);
      }

      bool operator<(const value_t& that) const noexcept {
        if (type_ != that.type_) {
          return type_ < that.type_;
        }
        switch (type_) {
          case type_t::nil:
            return false;
          case type_t::boolean:
            return boolean_ < that.boolean_;
          case type_t::number:
            return number_ < that.number_;
          case type_t::string:
            return *string_ < *that.string_;
          case type_t::table:
            return table_ < that.table_;
          case type_t::function:
            return function_ < that.function_;
        }
      }

    private:
      void construct(const value_t& that) noexcept {
        type_ = that.type_;
        switch (type_) {
          case type_t::nil:
            break;
          case type_t::boolean:
            boolean_ = that.boolean_;
            break;
          case type_t::number:
            number_ = that.number_;
            break;
          case type_t::string:
            new (&string_) string_ptr(that.string_);
            break;
          case type_t::table:
            new (&table_) table_ptr(that.table_);
            break;
          case type_t::function:
            new (&function_) function_ptr(that.function_);
            break;
        }
      }

      void construct(value_t&& that) noexcept {
        type_ = that.type_;
        that.type_ = type_t::nil;
        switch (type_) {
          case type_t::nil:
            break;
          case type_t::boolean:
            boolean_ = that.boolean_;
            break;
          case type_t::number:
            number_ = that.number_;
            break;
          case type_t::string:
            new (&string_) string_ptr(std::move(that.string_));
            break;
          case type_t::table:
            new (&table_) table_ptr(std::move(that.table_));
            break;
          case type_t::function:
            new (&function_) function_ptr(std::move(that.function_));
            break;
        }
      }

      void destruct() noexcept {
        switch (type_) {
          case type_t::nil:
            break;
          case type_t::boolean:
            break;
          case type_t::number:
            break;
          case type_t::string:
            string_.~shared_ptr();
            break;
          case type_t::table:
            table_.~shared_ptr();
            break;
          case type_t::function:
            function_.~shared_ptr();
            break;
        }
      }

      type_t type_;
      union {
        bool boolean_;
        double number_;
        string_ptr string_;
        table_ptr table_;
        function_ptr function_;
      };
    };

    inline value_t boolean(bool boolean) {
      value_t self;
      self.type_ = type_t::boolean;
      self.boolean_ = boolean;
      return self;
    }

    inline value_t number(double number) {
        value_t self;
        self.type_ = type_t::number;
        self.number_ = number;
      return self;
    }

    inline value_t string(const char* data, std::size_t size) {
      value_t self;
      self.type_ = type_t::string;
      new (&self.string_) string_ptr(std::make_shared<string_t>(data, size));
      return self;
    }

    inline value_t string(const std::string& data) {
      value_t self;
      self.type_ = type_t::string;
      new (&self.string_) string_ptr(std::make_shared<string_t>(data));
      return self;
    }

    inline value_t table() {
      value_t self;
      self.type_ = type_t::table;
      new (&self.table_) table_ptr(std::make_shared<table_t>());
      return self;
    }

    template <class T>
    inline value_t function(std::size_t argc, bool vararg, T&& closure) {
      value_t self;
      self.type_ = type_t::function;
      new (&self.function_) function_ptr(std::make_shared<function_t>(argc, vararg, std::forward<T>(closure)));
      return self;
    }

    inline const value_t& nil() noexcept {
      static const value_t self;
      return self;
    }

    inline const value_t& false_() noexcept {
      static const value_t self = boolean(false);
      return self;
    }

    inline const value_t& true_() noexcept {
      static const value_t self = boolean(true);
      return self;
    }

    inline array_ptr newarray(std::initializer_list<value_t> values, array_ptr array) {
      array_ptr result = std::make_shared<array_t>();
      for (const auto& value : values) {
        result->push_back(value);
      }
      if (array) {
        for (const auto& value : *array) {
          result->push_back(value);
        }
      }
      return result;
    }

    inline array_ptr newarray2(const value_t& value, array_ptr array) {
      array_ptr result = std::make_shared<array_t>();
      result->push_back(value);
      if (array) {
        for (const auto& value : *array) {
          result->push_back(value);
        }
      }
      return result;
    }

    inline const value_t& table_t::getmetafield(const char* event) const noexcept {
      if (metatable_) {
        const auto i = metatable_->map_.find(string(event));
        if (i != metatable_->map_.end()) {
          return i->second;
        }
      }
      return nil();
    }

    inline array_ptr function_t::call(array_ptr array) const {
      array_ptr A;
      array_ptr V;
      std::size_t i = 0;
      if (argc_) {
        A = std::make_shared<array_t>(argc_);
        for (; i < argc_; ++i) {
          (*A)[i] = get(array, i);
        }
      }
      if (vararg_) {
        V = std::make_shared<array_t>();
        for (; i < array->size(); ++i) {
          V->push_back(get(array, i));
        }
      }
      return closure_(A, V);
    }

    void value_t::setlist(double index, const value_t& value) const {
      if (!is_table()) {
        throw error_t("table expected");
      }
      // not care metafield because setlist is used from tableconstructor
      if (value.is_nil()) {
        table_->map_.erase(number(index));
      } else {
        table_->map_[number(index)] = value;
      }
    }

    void value_t::setlist(double index, array_ptr array) const {
      if (!is_table()) {
        throw error_t("table expected");
      }
      // not care metafield because setlist is used from tableconstructor
      for (const auto& value : *array) {
        if (value.is_nil()) {
          table_->map_.erase(number(index));
        } else {
          table_->map_[number(index)] = value;
        }
        ++index;
      }
    }

    value_t value_t::len() const {
      if (!is_table()) {
        throw error_t("table expected");
      }
      for (double i = 1; ; ++i) {
        if (gettable(number(i)).is_nil()) {
          return number(i - 1);
        }
      }
    }

    inline value_t open_env() {
      value_t env = table();

      env.settable(string("_G"), env);
      env.settable(string("_VERSION"), string("Lua 5.3"));

      env.settable(string("tonumber"), function(1, false, [](array_ptr A, array_ptr) -> array_ptr {
        return newarray({ number(get(A, 0).tonumber()) });
      }));

      env.settable(string("tostring"), function(1, false, [](array_ptr A, array_ptr) -> array_ptr {
        return newarray({ string(get(A, 0).tostring()) });
      }));

      env.settable(string("print"), function(0, true, [](array_ptr, array_ptr V) -> array_ptr {
        std::size_t i = 0;
        for (const auto& value : *V) {
          if (i > 0) {
            std::cout << "\t";
          }
          std::cout << value.tostring();
          ++i;
        }
        std::cout << "\n";
        return nullptr;
      }));

      env.settable(string("type"), function(1, false, [](array_ptr A, array_ptr) -> array_ptr {
        return newarray({ string(get(A, 0).type()) });
      }));

      return env;
    }
  }
}

#endif
]]

File: runtime/runtime_cxx.cpp
// Copyright (C) 2018 Tomoyuki Fujimori <moyu@dromozoa.com>
//
// This file is part of dromozoa-compiler.
//
// dromozoa-compiler is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// dromozoa-compiler is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// Under Section 7 of GPL version 3, you are granted additional
// permissions described in the GCC Runtime Library Exception, version
// 3.1, as published by the Free Software Foundation.
//
// You should have received a copy of the GNU General Public License
// and a copy of the GCC Runtime Library Exception along with
// dromozoa-compiler.  If not, see <http://www.gnu.org/licenses/>.

#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <utility>

#include "runtime_cxx_value.hpp"

namespace dromozoa {
  namespace runtime {
    namespace {
      void copy_construct(value_t& self, const value_t& that) {
        self.type = that.type;
        switch (self.type) {
          case type_t::nil:
            break;
          case type_t::boolean:
            self.boolean = that.boolean;
            break;
          case type_t::number:
            self.number = that.number;
            break;
          case type_t::string:
            new (&self.string) string_ptr(that.string);
            break;
          case type_t::table:
            new (&self.table) table_ptr(that.table);
            break;
          case type_t::function:
            new (&self.function) function_ptr(that.function);
            break;
          default:
            throw std::runtime_error("unreachable code");
        }
      }

      void move_construct(value_t& self, value_t&& that) {
        self.type = that.type;
        that.type = type_t::nil;
        switch (self.type) {
          case type_t::nil:
            break;
          case type_t::boolean:
            self.boolean = that.boolean;
            break;
          case type_t::number:
            self.number = that.number;
            break;
          case type_t::string:
            new (&self.string) string_ptr(std::move(that.string));
            break;
          case type_t::table:
            new (&self.table) table_ptr(std::move(that.table));
            break;
          case type_t::function:
            new (&self.function) function_ptr(std::move(that.function));
            break;
          default:
            throw std::runtime_error("unreachable code");
        }
      }

      void destruct(value_t& self) {
        switch (self.type) {
          case type_t::nil:
            break;
          case type_t::boolean:
            break;
          case type_t::number:
            break;
          case type_t::string:
            self.string.~shared_ptr();
            break;
          case type_t::table:
            self.table.~shared_ptr();
            break;
          case type_t::function:
            self.function.~shared_ptr();
            break;
          default:
            throw std::runtime_error("unreachable code");
        }
      }

      struct noop_t : function_t {
        virtual array_t operator()(array_t) {
          return {};
        }
      };
    }

    value_t NIL = type_t::nil;
    value_t FALSE = false;
    value_t TRUE = true;
    value_t string_metatable = type_t::table;

    value_t::value_t()
      : mode(mode_t::variable),
        type(type_t::nil) {}

    value_t::value_t(const value_t& that)
      : value_t() {
      copy_construct(*this, that);
    }

    value_t::value_t(value_t&& that)
      : value_t() {
      move_construct(*this, std::move(that));
    }

    value_t::~value_t() {
      destruct(*this);
    }

    value_t& value_t::operator=(const value_t& that) {
      if (mode == mode_t::constant) {
        throw std::runtime_error("cannot assign to constant value");
      }
      destruct(*this);
      copy_construct(*this, that);
      return *this;
    }

    value_t& value_t::operator=(value_t&& that) {
      if (mode == mode_t::constant) {
        throw std::runtime_error("cannot assign to constant value");
      }
      destruct(*this);
      move_construct(*this, std::move(that));
      return *this;
    }

    value_t::value_t(type_t type)
      : mode(mode_t::constant),
        type(type) {
      switch (type) {
        case type_t::nil:
          break;
        case type_t::boolean:
          boolean = false;
          break;
        case type_t::number:
          number = 0;
          break;
        case type_t::string:
          new (&string) string_ptr(std::make_shared<std::string>());
          break;
        case type_t::table:
          new (&table) table_ptr(std::make_shared<table_t>());
          break;
        case type_t::function:
          new (&function) function_ptr(std::make_shared<noop_t>());
          break;
        default:
          throw std::runtime_error("unreachable code");
      }
    }

    value_t::value_t(bool boolean)
      : mode(mode_t::constant),
        type(type_t::boolean) {
      this->boolean = boolean;
    }

    value_t::value_t(double number)
      : mode(mode_t::constant),
        type(type_t::number) {
      this->number = number;
    }

    value_t::value_t(const char* data)
      : mode(mode_t::constant),
        type(type_t::string) {
      new (&this->string) string_ptr(std::make_shared<std::string>(data));
    }

    value_t::value_t(const char* data, std::size_t size)
      : mode(mode_t::constant),
        type(type_t::string) {
      new (&this->string) string_ptr(std::make_shared<std::string>(data, size));
    }

    value_t::value_t(const std::string& string)
      : mode(mode_t::constant),
        type(type_t::string) {
      new (&this->string) string_ptr(std::make_shared<std::string>(string));
    }

    value_t::value_t(std::string&& string)
      : mode(mode_t::constant),
        type(type_t::string) {
      new (&this->string) string_ptr(std::make_shared<std::string>(std::move(string)));
    }

    bool value_t::operator<(const value_t& that) const {
      if (type != that.type) {
        return type < that.type;
      }
      switch (type) {
        case type_t::nil:
          return false;
        case type_t::boolean:
          return boolean < that.boolean;
        case type_t::number:
          return number < that.number;
        case type_t::string:
          return string < that.string;
        case type_t::table:
          return table < that.table;
        case type_t::function:
          return function < that.function;
        default:
          throw std::runtime_error("unreachable code");
      }
    }

    bool value_t::is_nil() const {
      return type == type_t::nil;
    }

    bool value_t::is_boolean() const {
      return type == type_t::boolean;
    }

    bool value_t::is_number() const {
      return type == type_t::number;
    }

    bool value_t::is_string() const {
      return type == type_t::string;
    }

    bool value_t::is_table() const {
      return type == type_t::table;
    }

    bool value_t::is_function() const {
      return type == type_t::function;
    }

    bool value_t::is_false() const {
      return is_boolean() && !boolean;
    }

    bool value_t::is_true() const {
      return is_boolean() && boolean;
    }

    bool value_t::is_nil_or_false() const {
      return is_nil() || is_false();
    }

    array_t::array_t()
      : size() {}

    array_t::array_t(std::size_t n)
      : array_t() {
      if (n > 0) {
        data = std::shared_ptr<value_t>(new value_t[n], std::default_delete<value_t[]>());
        size = n;
      }
    }

    array_t::array_t(std::initializer_list<value_t> source)
      : array_t() {
      const auto n = source.size();
      if (n > 0) {
        data = std::shared_ptr<value_t>(new value_t[n], std::default_delete<value_t[]>());
        size = n;
        auto* ptr = data.get();
        for (const auto& value : source) {
          *ptr++ = value;
        }
      }
    }

    array_t::array_t(const value_t& value, array_t array) {
      const auto n = 1 + array.size;
      if (n > 0) {
        data = std::shared_ptr<value_t>(new value_t[n], std::default_delete<value_t[]>());
        size = n;
        auto* ptr = data.get();
        *ptr++ = value;
        for (std::size_t i = 0; i < array.size; ++i) {
          *ptr++ = array.data.get()[i];
        }
      }
    }

    value_t& array_t::operator[](std::size_t i) const {
      if (i < size) {
        return data.get()[i];
      } else {
        return NIL;
      }
    }

    array_t array_t::sub(std::size_t begin) const {
      if (size > begin) {
        array_t that(size - begin);
        auto* ptr = that.data.get();
        for (std::size_t i = begin; i < size; ++i) {
          *ptr++ = data.get()[i];
        }
        return that;
      } else {
        return {};
      }
    }

    array_t array_t::sub(std::size_t begin, std::size_t end) const {
      if (end > begin) {
        array_t that(end - begin);
        auto* ptr = that.data.get();
        for (std::size_t i = begin; i < end; ++i) {
          *ptr++ = (*this)[i];
        }
        return that;
      } else {
        return {};
      }
    }

    const value_t& rawget(const value_t& self, const value_t& index) {
      if (!self.is_table()) {
        throw value_t("bad argument #1");
      }
      const auto i = self.table->map.find(index);
      if (i == self.table->map.end()) {
        return NIL;
      } else {
        return i->second;
      }
    }

    const value_t& rawset(const value_t& self, const value_t& index, const value_t& value) {
      if (!self.is_table()) {
        throw value_t("bad argument #1");
      }
      if (index.is_nil()) {
        throw value_t("table index is nil");
      }
      if (value.is_nil()) {
        self.table->map.erase(index);
      } else {
        self.table->map[index] = value;
      }
      return self;
    }


    const value_t& getmetafield(const value_t& self, const value_t& event) {
      value_t metatable;
      if (self.is_string()) {
        metatable = string_metatable;
      } else if (self.is_table()) {
        metatable = self.table->metatable;
      }
      if (metatable.is_table()) {
        return rawget(metatable, event);
      } else {
        return NIL;
      }
    }

    std::string type(const value_t& self) {
      switch (self.type) {
        case type_t::nil:
          return "nil";
        case type_t::boolean:
          return "boolean";
        case type_t::number:
          return "number";
        case type_t::string:
          return "string";
        case type_t::table:
          return "table";
        case type_t::function:
          return "function";
        default:
          throw std::runtime_error("unreachable code");
      }
    }

    std::string tostring(const value_t& self) {
      switch (self.type) {
        case type_t::nil:
          return "nil";
        case type_t::boolean:
          if (self.boolean) {
            return "true";
          } else {
            return "false";
          }
        case type_t::number:
          {
            std::ostringstream out;
            out << std::setprecision(17) << self.number;
            return out.str();
          }
        case type_t::string:
          return *self.string;
        case type_t::table:
          {
            // TODO metatable
            const auto& field = getmetafield(self, "__tostring");
            if (!field.is_nil()) {
              // return field.call1(*this);
              return "__";
            } else {
              std::ostringstream out;
              out << "table: " << self.table.get();
              return out.str();
            }
          }
        case type_t::function:
          {
            std::ostringstream out;
            out << "function: " << self.function.get();
            return out.str();
          }
        default:
          throw std::runtime_error("unreachable code");
      }
    }

    array_t call(const value_t& self, const array_t& args) {
      if (self.is_function()) {
        return (*self.function)(args);
      } else {
        const auto& field = getmetafield(self, "__call");
        if (field.is_function()) {
          array_t new_args(args.size + 1);
          return (*field.function)(array_t(field, args));
        } else {
          throw value_t("attempt to call a " + type(self) + " value");
        }
      }
    }

    value_t call1(const value_t& self, const array_t& args) {
      return call(self, args)[0];
    }

    void call0(const value_t& self, const array_t& args) {
      call(self, args);
    }
  }
}
